addon/components/frost-scroll.js

```diff
@@ -70,7 +70,9 @@ export default Component.extend({
     }
 
     this._mouseEnterHandler = () => {
-      this.get('onMouseEnter')(this.get('element'))
+      run(() => {
```

Comment:
This does not feel like the right way to solve anything.  The setting of this handler via `this.$().on('mouseenter', this._mouseEnterHandler)` is already scheduled via `run.scheduleOnce('afterRender',` and when the `mouseenter` event is being fired there is no run loop related activity occurring (meaning a mouse over occurred - there's not rerendering etc going on).   Have you exhausted the usage of this component in the application for which the tests are failing as well as the test setup?

Reply:
I haven't checked the _addon/components/frost-select-dropdown.js_ file changes yet but suspect my comments will be the same

Reply:
Isn't it possible that the consumer of this addon would do `set` and `setProperties` inside of their `onMouseEnter` callback. If it wasn't inside a run loop, every `set` would trigger a render.

Reply:
I took this one out, I don't see any usage's in our app so it doesn't really matter for us. 
The frost-select `mousedown`, `mouseenter`, and `_keyDownHandler` are 100% a problem tho. If I turn off the auto run loop i can reproduce the testing error consistently we intermittently see our app that interact with `frost-select-dropown`

Reply:
@sophypal That is certainly a possibility though that onus should be on the application, as they are choosing to do things in their callback that is causing this issue.  If they chose not to do this at all, or did not update properties that would cause problems this problem would not exist.  And the properties they are choosing or not choosing to update are properties in their application code (or consumption therein) not any properties provided by this component, which again, makes it an application concern.

Reply:
If the application was consuming a 3rd-party non-ember library, then yes the onus should be on them. However, they're importing an Ember addon which needs to be safe to use. Why would you expect all consumers of an "Ember addon" to assume that any callbacks you pass won't be async safe? Shouldn't the responsibility be on the addon for managing this compliance when it was the addon that imported the 3rd-party lib. How are consumers to know, without peeking through the code, that it was using native events?

Reply:
Using your logic @notmessenger, you're saying that any callback I pass to an Ember addon should be surrounded with `run`, if I need to set my local properties. That seems wrong.

Reply:
It's not this add-on causing the problem though.  The problem is that the consuming application is attempting to modifying properties multiple times in the same render.  As we've discussed before the way to solve this is to adjust the code flow triggering computed properties, not to wrap everything in a run loop.  The point is this can be solved in the consuming application - not by using a run loop but by orchestrating a proper code flow.  This makes this issue not one of where the run loop logic should go - and that it should be in the add-on for all to use and not have think about implementing - but one of how to consume this functionality correctly.  The fact that we're changing this add-on to run changes to the consuming apps code at a different time to avoid changing its value in the same render means that its the app that's not orchestrated correctly.

Reply:
And even if I were to agree that this should be done in the add-on, we're saying that `run()` is good enough scheduling?  What about all of the available run queues?  How are those to be coordinated with an application's needs?  They wouldn't be - it should be up to the app to handle the use of `run()`, and more correctly, to not need to use it at all.

Reply:
I understand your concern but it's actually pretty common that applications set properties during an action caused by some event.  The issue you're describing is when applications modify properties multiple times during a single render.  I don't see how a mouse event being handled would cause a double render if the event happens outside the render process.  Anyways, we maybe saying the same thing but not understanding each other.  I just find that generally, it's a good practice to guard event handling code that is not run loop aware in a run loop.

Reply:
@notmessenger side note:
`Ember.run(...)` doesn't avoid the same render error. `Ember.run(..)`executes the callback immediately. You're thinking of `Ember.run.next(...)` that'll wait for the next render.
https://ember-twiddle.com/b9b2d1a1bb13a0ad86e202d0db3e57f1?openFiles=components.my-component.js%2Ctemplates.components.my-component.hbs

Reply:
That's an excellent point and a point I was trying to make as well. The use of a `run` loop inside an non-ember async event is to avoid unexpected async behavior not necessarily to avoid the render issue were we seeing scarcely in mcp. I totally agree with @notmessenger in that using `run.next` or `run.schedule` as a means to avoid the render error is not recommended but I didn't think @EWhite613 added it to avoid the render issues.

Reply:
Except you previously said @sophypal 

> If it wasn't inside a run loop, every set would trigger a render if it altered properties that affected computed properties it needed to render.

and @EWhite613 agreed, meaning that @EWhite613 added it to avoid the render issues.

Reply:
@notmessenger I added it to avoid users from failing when the autorun loop is disabled. I used `run` not `run.next`, render issues would still be prevalent

Reply:
LOL. I did say that didn't I. I can see the misunderstanding now. When I said that, I didn't mean the specific multiple render error we've been seeing since the upgrade to Ember 2.12. I've come across issues, related to rendering, that are triggered due to 2 separate async events executing at similar timeframes. I incorrectly implied that using a run block would resolve the two events into one render, whilst not using a run block would trigger multiple renders--which is not the same as the double render error. Using `run.schedule`, will allow parallel stream of events to coalesce into one render call if they happen to execute before the next scheduled queue runs.

There are other unpredictable behaviors, but the issues I've seen in the past are observed as rendering issues (i.e. doing `set` outside a run loop), and they most often break our test because the auto run loop is disabled.

Reply:
To clarify even more, the double render error I mentioned is errors due to setting properties during a render event which is more like a cascade rendering issue. The other rendering issue I discussed occur as 2 separate rendering event due to 2 separate async events.